### ISSUE TITLE
"FATAL ERROR: invalid table size Allocation failed" with many inferred types

### ISSUE THREAD

**Issue by ProdigySim** [microsoft/TypeScript#31420]
**TypeScript Version:**  
3.3.3333
3.3.4000
3.4.5
Version 3.5.0-dev.20190515

**Search Terms:**
- "invalid table size": this one is new
- "Allocation Failed"

**Code**
I will continue trying to build a reduced repro, but the code right now is a private React-Native project with Redux. We have ~200 source TS files and ~70 basic redux actions, as well as some [thunk actions](https://github.com/reduxjs/redux-thunk).

We use a lot of type inference to build a discriminated type union of our redux types, similar to [this article](https://medium.com/@martin_hotell/improved-redux-type-safety-with-typescript-2-8-2c11a8062575). **Right now, if I add one additional redux type to our application, it will crash with the below Allocation Failed error**

Here is a reasonable facsimile of one of our action files:
```ts
// utility helper to get inference of the correct action type (TODO replace w/ `as const`)
export function createAction<T extends string, U extends Action<T>>(action: U) {
  return action;
}

function myReduxActionOne() {
  return createAction({ type: 'some_long_string_constant' });
}
function myReduxActionTwo(data: object) {
  return createAction({ 
    type: 'some_long_string_constant',
    data
  });
}

export type MyReduxActions = ReturnType<
  | typeof myReduxActionOne
  | typeof myReduxActionTwo
  // If I add one more `typeof actionCreator` to this union, 
  // the program will crash with OOM.
>;
```
These action types are used in segmented reducers, and eventually get combined into one large discriminated union of types near a redux `combineReducers()` call. I can provide more sample code, and will continue to work on a reduced repro.

**Expected behavior:**
`tsc` compiles the codebase without error.

**Actual behavior:**
`tsc` crashes with a unique out of memory error:
```
$ tsc --noEmit --diagnostics

<--- Last few GCs --->

[16153:0x102645000]   147934 ms: Scavenge 2141.2 (2240.4) -> 2126.5 (2240.9) MB, 9.1 / 0.0 ms  (average mu = 0.998, current mu = 1.000) allocation failure 
[16153:0x102645000]   148047 ms: Scavenge 2141.7 (2240.9) -> 2127.0 (2241.4) MB, 8.7 / 0.0 ms  (average mu = 0.998, current mu = 1.000) allocation failure 
[16153:0x102645000]   148160 ms: Scavenge 2142.2 (2241.4) -> 2127.5 (2241.9) MB, 15.1 / 0.0 ms  (average mu = 0.998, current mu = 1.000) allocation failure 


<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0x2c389e8dbe3d]
Security context: 0x1fe24991e6e1 <JSObject>
    1: set [0x1fe249911ea1](this=0x1fe2f27d91d9 <Map map = 0x1fe2a7104359>,0x1fe2ac0b0e59 <String[11]: 545159,4710>,2)
    2: recursiveTypeRelatedTo(aka recursiveTypeRelatedTo) [0x1fe2bf802201] [/Users/mbusby/snip-project-name/node_modules/typescript/lib/tsc.js:~35308] [pc=0x2c389e8f82e5](this=0x1fe23dd826f1 <undefined>,source=0x1fe2a9be4a09 <Type map = 0x1fe2...

FATAL ERROR: invalid table size Allocation failed - JavaScript heap out of memory
 1: 0x10003b125 node::Abort() [/Users/mbusby/.nvm/versions/v10.15.0/bin/node]
 2: 0x10003b32f node::OnFatalError(char const*, char const*) [/Users/mbusby/.nvm/versions/v10.15.0/bin/node]
 3: 0x1001a8e85 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/Users/mbusby/.nvm/versions/v10.15.0/bin/node]
 4: 0x1005742a2 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/Users/mbusby/.nvm/versions/v10.15.0/bin/node]
 5: 0x1006e0756 v8::internal::OrderedHashTable<v8::internal::OrderedHashMap, 2>::Rehash(v8::internal::Handle<v8::internal::OrderedHashMap>, int) [/Users/mbusby/.nvm/versions/v10.15.0/bin/node]
 6: 0x1007a8793 v8::internal::Runtime_MapGrow(int, v8::internal::Object**, v8::internal::Isolate*) [/Users/mbusby/.nvm/versions/v10.15.0/bin/node]
 7: 0x2c389e8dbe3d 
```

**Playground Link:** None yet

**Related Issues:** #31281  looks similar, but the error message was very different.

**Other diagnostic info**

If I remove the 1 action that puts `tsc` over the edge, I can get diagnostic info on various versions of typescript:

On 3.3.3333:
```
Files:            644
Lines:         352690
Nodes:        1075014
Identifiers:   391809
Symbols:       505217
Types:         490478
Memory used: 3475902K
I/O read:       0.15s
I/O write:      0.00s
Parse time:     2.94s
Bind time:      1.13s
Check time:   369.67s
Emit time:      0.00s
Total time:   373.73s
```

On Version 3.5.0-dev.20190515:
```
$ tsc --noEmit --diagnostics
Files:            644
Lines:         353393
Nodes:        1075852
Identifiers:   391910
Symbols:       573963
Types:         537306
Memory used: 3537350K
I/O read:       0.15s
I/O write:      0.00s
Parse time:     2.81s
Bind time:      1.14s
Check time:   262.53s
Emit time:      0.00s
Total time:   266.47s
```
These numbers were collected with a tsconfig that looks like:
```json
{
    "compilerOptions": {
      "esModuleInterop": true,
      "allowJs": false,
      "target": "es6",
      "lib": ["dom", "es2018"],
      "jsx": "react-native",
      "alwaysStrict": true,
      "strict": false,
      "module": "esnext",
      "moduleResolution": "node",
      "skipLibCheck": true,
      "skipDefaultLibCheck": true,
      "allowSyntheticDefaultImports": true,
      "baseUrl": "./",
      "types": [
        "jest",
        "node"
      ]
    },
    "include": [
      "src"
    ],
    "exclude": [
      "node_modules",
      "android",
      "ios"
    ]
  }
```

One thing I noticed in comparison to some of my other similar-but-larger projects is that this has a very large number of "Types" in the diagnostic info.  This project shows 500k types in 650 files vs  the ~80k types in 2000 files I have in another project.

**Comment by ProdigySim** [comment]
I think the issue is related to my inference of action types through reducers. 

```ts
// pseudocode of reducers:
declare function fooReducer(state: IFooState | undefined, action: MyFooReduxActions): IFooState;
declare function barReducer(state: IBarState | undefined, action: MyBarReduxActions): IBarState;  

const appReducers = {
  foo: fooReducer,
  bar: barReducer,
};

// Read computed reducer type
type TAppReducer = typeof appReducers;
// Build the type of our application state through the return types of all reducers
export type TAppState = {
  [key in keyof TAppReducer]: ReturnType<TAppReducer[key]>
};
// Build the discriminated union of all our actions using inference
type TAppActions = Values<TAppReducer> extends Reducer<any, infer TActions> ? TActions : never;
// Type our final reducer using these values
type TRootReducer = Reducer<TAppState, TAppActions>;

// Build the root reducer
export const rootReducer: TRootReducer = combineReducers<TAppState, TAppActions>(appReducers);
```

If I remove all the type logic and just do `export const rootReducer = combineReducers(appReducers)`, it compiles without error, but the root reducer gets the type of `Reducer<TAppState, AnyAction>`, where AnyAction is overly generic. Also the number of "Types" in the `--diagnostics` info goes down to 90230 (90% reduction). So I think I found the culprit :(

Would love to build all my redux types through inference like this, but maybe I will have to manually collect action & state types like a caveman.

**Comment by ProdigySim** [comment]
I figured out the root of my problem, reduceable to a Typescript Playground. This is primarily error on my part.

[Demo of the core issue that was causing a type bomb here](http://www.typescriptlang.org/play/#src=type%20F1%20%3D%20(x%3A%20%7B%20y%3A%20'a'%20%7D%20%7C%20%7B%20y%3A%20'b'%20%7D%20%7C%20%7By%3A%20'c'%20%7D)%20%3D%3E%20void%3B%0D%0Atype%20F2%20%3D%20(x%3A%20%7B%20y%3A%20'd'%20%7D%20%7C%20%7B%20y%3A%20'e'%20%7D%20%7C%20%7By%3A%20'f'%20%7D)%20%3D%3E%20void%3B%0D%0Atype%20FX%3CT%3E%20%3D%20(x%3A%20T)%20%3D%3E%20void%3B%0D%0A%0D%0A%0D%0Atype%20x%20%3D%20F1%20extends%20(x%3A%20infer%20U)%3D%3Evoid%20%3F%20U%20%3A%20never%3B%0D%0A%0D%0Atype%20AllFuncs%20%3D%20F1%20%7C%20F2%3B%0D%0A%0D%0A%2F%2F%20Should%20be%3A%0D%0A%2F%2F%20%7B%20y%3A%20'a'%20%7D%20%7C%20%7B%20y%3A%20'b'%20%7D%20%7C%20%7B%20y%3A%20'c'%20%7D%20%7C%20...%0D%0A%2F%2F%20Is%20actually%3A%0D%0A%2F%2F%20(%7B%20y%3A%20%22a%22%3B%20%7D%20%26%20%7B%20y%3A%20%22d%22%3B%20%7D)%20%7C%20(%7B%20y%3A%20%22a%22%3B%20%7D%20%26%20%7B%20y%3A%20%22e%22%3B%20%7D)%20%7C%20...%0D%0Atype%20AllParams%20%3D%20AllFuncs%20extends%20FX%3Cinfer%20U%3E%20%3F%20U%20%3A%20never%3B%0D%0A%0D%0A%0D%0A%0D%0A%2F%2F%20Direct%20PoC%0D%0Afunction%20FF%3CT%2C%20U%3E()%20%7B%0D%0A%20%20type%20fns%20%3D%20((x%3A%20T)%20%3D%3E%20void)%20%7C%20((x%3A%20U)%20%3D%3E%20void)%3B%0D%0A%20%20%2F%2F%20xx%20%3D%20T%20%26%20U%0D%0A%20%20type%20xx%20%3D%20fns%20extends%20(x%3A%20infer%20Z)%20%3D%3E%20void%20%3F%20Z%20%3A%20never%0D%0A%7D)
Basically, I expected 
```ts
(((x: T) => void) | ((x: U) => void)) extends (x: infer V) => void ? V : never
```
to resolve to `T | U`. But instead it resolves to `T & U`. Which makes sense conceptually--I had just expected the "extends" test to map across the union.

So, instead of `n1 + n2 + n3` types of actions, I produced `n1 * n2 * n3` types of actions... which grew to use a ton of memory.

The only oddity left is that a type like `type x = { y: 'a' } & { y: 'b' };` can be considered valid. But I'd guess this is probably captured in some other issue report.
